src/lib/utils.py

```diff
@@ -108,13 +114,118 @@ def grouped_transform(
     return data
 
 
-def grouped_diff(data: DataFrame, keys: List[str], skip: List[str] = None) -> DataFrame:
+def grouped_diff(
+    data: DataFrame,
+    keys: List[str],
+    skip: List[str] = None,
+    prefix: Tuple[str, str] = ("new_", "total_"),
+    compat: bool = True,
+) -> DataFrame:
     return grouped_transform(
-        data, keys, lambda x: x.ffill().diff(), skip=skip, prefix=("new_", "total_")
+        data, keys, lambda x: x.ffill().diff(), skip=skip, prefix=prefix, compat=compat
     )
 
 
-def grouped_cumsum(data: DataFrame, keys: List[str], skip: List[str] = None) -> DataFrame:
+def grouped_cumsum(
+    data: DataFrame,
+    keys: List[str],
+    skip: List[str] = None,
+    prefix: Tuple[str, str] = ("total_", "new_"),
+    compat: bool = True,
+) -> DataFrame:
     return grouped_transform(
-        data, keys, lambda x: x.fillna(0).cumsum(), skip=skip, prefix=("total_", "new_")
+        data, keys, lambda x: x.fillna(0).cumsum(), skip=skip, prefix=prefix, compat=compat
+    )
+
+
+def stack_table(
+    data: DataFrame, index_columns: List[str], value_columns: List[str], stack_columns: List[str]
+) -> DataFrame:
+    """
+    Pivots a DataFrame's columns and aggregates the result as new columns with suffix. E.g.:
+
+    data:
+
+    idx piv val
+     0   A   1
+     0   B   2
+     1   A   3
+     1   B   4
+
+    stack_table(data, index_columns=[idx], value_columns=[val], stack_columns=[piv]):
+
+    idx val val_A val_B
+     0   3    1     2
+     1   7    3     4
+    """
+    output = data.drop(columns=stack_columns).groupby(index_columns).sum()
+
+    for col_stack in stack_columns:
+        col_stack_values = data[col_stack].dropna().unique()
+        for col_variable in value_columns:
+            df = data[index_columns + [col_variable, col_stack]].copy()
+            df = df.pivot_table(
+                values=col_variable, index=index_columns, columns=[col_stack], aggfunc="sum"
+            )
+            column_mapping = {suffix: f"{col_variable}_{suffix}" for suffix in col_stack_values}
+            df = df.rename(columns=column_mapping)
+            transfer_columns = list(column_mapping.values())
+            output[transfer_columns] = df[transfer_columns]
+
+    return output.reset_index()
+
+
+def age_group(age: int) -> str:
+    """
+    Categorical age group given a specific age, codified into a function to enforce consistency.
+    """
+    if age < 15:
+        return "children"
+    elif age < 65:
+        return "adult"
+    else:
+        return "elderly"
+
+
+def process_records(data: DataFrame, metadata: DataFrame) -> DataFrame:
```

Comment:
Please add a docstring -- I knew what this did because of your description in the pull request, but otherwise I'd have to read the code to know what it did.

Reply:
It would also help to use a more descriptive name, maybe:
infer_new_or_total
infer_new_and_total
infer_new_or_total_columns
infer_new_and_total_columns

Reply:
This is a great point. I decided to refactor this function and break it up into multiple, better documented methods. PTAL.